File: backend/src/services/configurator/quotationCompiler.js
'use strict';

/**
 * quotationCompiler.js — pure orchestration layer.
 *
 * Combines:
 *   • bomEngine.expandConfig(...)
 *   • pricingEngine.computeQuote(...)
 *   • labourEngine.aggregateHoursFromBomRows(...)
 *
 * into a single deterministic object that mirrors the structure
 * persisted into `configurator_quotations` (`bom_spec` + `pricing_spec`
 * JSONB columns) and consumed by the PDF generator.
 *
 * NO database / network access here. The caller must hydrate the
 * components catalog map and pass it in.
 */

const bomEngine = require('./bomEngine');
const pricingEngine = require('./pricingEngine');
const labourEngine = require('./labourEngine');

const num = (v, fallback = 0) => (typeof v === 'number' && Number.isFinite(v) ? v : fallback);

/**
 * Build a fully-evaluated quotation payload.
 *
 * @param {object}   args
 * @param {object}   args.configuration                Sequelize plain object (or any { id, code, name, ... })
 * @param {object}   args.configData                   The Configuration.config_data blob
 * @param {object}   args.catalog                      { byId, byPartNumber } from configuratorService
 * @param {object}   args.lookup                       LookupRates (LBR_*_rate, OVERHEAD_PCT, COPPER_RATE_PER_LB)
 * @param {object}   args.pricing                      PricingStrategy
 * @param {object}   [args.schedule]                   ScheduleInput
 * @param {string[]} [args.holidays]                   ISO YYYY-MM-DD
 * @param {object[]} [args.lineAdders]                 Cross-section adders [{section_id, desc, value}]
 * @param {object[]} [args.preBuiltSections]           Optional override — bypass BOM-derived sections
 * @returns {object} compiled quotation
 */
function compileQuotation({
  configuration,
  configData,
  catalog,
  lookup,
  pricing,
  schedule = {},
  holidays = [],
  lineAdders = [],
  preBuiltSections = null,
}) {
  if (!lookup) throw new Error('compileQuotation: lookup is required');
  if (!pricing) throw new Error('compileQuotation: pricing is required');

  const bom = bomEngine.expandConfig(configData || {}, catalog || { byId: new Map(), byPartNumber: new Map() });

  // Derive sections — either preBuilt (manual) or from BOM rows
  let sections;
  if (Array.isArray(preBuiltSections) && preBuiltSections.length) {
    sections = preBuiltSections;
  } else {
    sections = bomEngine.sectionsFromBomRows(bom.rows);
  }

  // Inject cross-section line-item adders
  if (lineAdders && lineAdders.length) {
    const bySec = new Map();
    for (const a of lineAdders) {
      const id = a.section_id != null ? String(a.section_id) : (sections[0] && sections[0].id) || '0';
      if (!bySec.has(id)) bySec.set(id, []);
      bySec.get(id).push({ desc: String(a.desc || ''), value: num(a.value, 0) });
    }
    for (const sec of sections) {
      const extras = bySec.get(String(sec.id));
      if (extras) sec.line_items = [...(sec.line_items || []), ...extras];
    }
  }

  const quote = pricingEngine.computeQuote({
    sections,
    lookup,
    pricing,
    schedule,
    holidays,
  });

  // Stand-alone labour roll-up (from raw BOM rows — useful for the
  // labour preview tab even when no sections are active yet).
  const labour = labourEngine.computeLabour(bom.rows, lookup);

  // Quotation items array → mirrors `configurator_quotation_items` rows.
  // We expose this so configuratorService can persist line-by-line.
  const items = bom.rows.map((r, idx) => ({
    line_number: idx + 1,
    component_id: r.component_id,
    part_number: r.part_number,
    description: r.description || r.name || '',
    category: r.category,
    step_key: r.step_key,
    section_number: r.section_number,
    quantity: r.quantity,
    unit: r.unit,
    unit_cost: r.unit_cost,
    total_cost: r.total_cost,
    meta: r.meta,
  }));

  // bom_spec / pricing_spec — the JSONB blobs persisted on the quotation.
  const bom_spec = {
    rows: bom.rows,
    by_step: bom.by_step,
    by_section: bom.by_section,
    totals: bom.totals,
  };
  const pricing_spec = {
    lookup,
    pricing_strategy: pricing,
    schedule,
    holidays,
    line_adders: lineAdders,
    sections_input: sections,
    quote,            // full pricing engine output
    labour_summary: labour,
  };

  return {
    configuration: configuration ? {
      id: configuration.id || null,
      code: configuration.code || null,
      name: configuration.name || null,
      project_id: configuration.project_id || null,
    } : null,
    quote,
    labour,
    items,
    bom_spec,
    pricing_spec,
    totals: {
      ...quote.totals,
      total_cost: quote.total_cost,
      target_price: quote.pricing.target_price,
      rounded_price: quote.pricing.rounded_price,
      actual_profit: quote.pricing.actual_profit,
      actual_gm: quote.pricing.actual_gm,
    },
  };
}

module.exports = {
  compileQuotation,
};

File: java-backend/src/main/java/com/forge/configurator/engine/QuotationCompiler.java
package com.forge.configurator.engine;

import java.util.*;

public class QuotationCompiler {
    private final BomEngine bomEngine;
    private final PricingEngine pricingEngine;
    private final LabourEngine labourEngine;

    public QuotationCompiler(BomEngine bomEngine, PricingEngine pricingEngine, LabourEngine labourEngine) {
        this.bomEngine = bomEngine;
        this.pricingEngine = pricingEngine;
        this.labourEngine = labourEngine;
    }

    public record LineAdder(String sectionId, String desc, double value) {
    }

    public record QuotationItem(int lineNumber,
                                UUID componentId,
                                String partNumber,
                                String description,
                                String category,
                                String stepKey,
                                Integer sectionNumber,
                                double quantity,
                                String unit,
                                double unitCost,
                                double totalCost,
                                Map<String, Object> meta) {
    }

    public record CompiledQuotation(Map<String, Object> configuration,
                                    PricingEngine.QuoteResult quote,
                                    LabourEngine.LabourSummary labour,
                                    List<QuotationItem> items,
                                    Map<String, Object> bomSpec,
                                    Map<String, Object> pricingSpec,
                                    Map<String, Double> totals) {
    }

    public CompiledQuotation compileQuotation(Map<String, Object> configuration,
                                              Map<String, Object> configData,
                                              BomEngine.Catalog catalog,
                                              Map<String, Double> lookup,
                                              PricingEngine.PricingStrategy pricing,
                                              PricingEngine.ScheduleInput schedule,
                                              List<java.time.LocalDate> holidays,
                                              List<LineAdder> lineAdders,
                                              List<PricingEngine.SectionInput> preBuiltSections) {
        BomEngine.ExpansionResult bom = bomEngine.expandConfig(configData, catalog);

        List<PricingEngine.SectionInput> sections;
        if (preBuiltSections != null && !preBuiltSections.isEmpty()) {
            sections = new ArrayList<>(preBuiltSections);
        } else {
            sections = new ArrayList<>(bomEngine.sectionsFromBomRows(bom.rows()));
        }

        if (lineAdders != null && !lineAdders.isEmpty()) {
            Map<String, List<PricingEngine.SectionLineItem>> grouped = new HashMap<>();
            for (LineAdder adder : lineAdders) {
                String id = adder.sectionId() == null || adder.sectionId().isBlank()
                        ? (sections.isEmpty() ? "0" : sections.getFirst().id())
                        : adder.sectionId();
                grouped.computeIfAbsent(id, key -> new ArrayList<>())
                        .add(new PricingEngine.SectionLineItem(adder.desc(), adder.value()));
            }

            sections = sections.stream().map(section -> {
                List<PricingEngine.SectionLineItem> extras = grouped.get(section.id());
                if (extras == null || extras.isEmpty()) {
                    return section;
                }
                List<PricingEngine.SectionLineItem> merged = new ArrayList<>(section.lineItems());
                merged.addAll(extras);
                return new PricingEngine.SectionInput(
                        section.id(),
                        section.description(),
                        section.unitMaterialCost(),
                        section.qty(),
                        section.laborHoursPerUnit(),
                        section.copperWeightPerUnit(),
                        merged
                );
            }).toList();
        }

        PricingEngine.QuoteResult quote = pricingEngine.computeQuote(
                new PricingEngine.QuoteInput(sections, lookup, pricing, schedule, holidays)
        );

        LabourEngine.LabourSummary labour = labourEngine.computeLabour(bom.rows(), lookup);

        List<QuotationItem> items = new ArrayList<>();
        for (int i = 0; i < bom.rows().size(); i++) {
            BomEngine.BomRow row = bom.rows().get(i);
            items.add(new QuotationItem(
                    i + 1,
                    row.componentId(),
                    row.partNumber(),
                    row.description() == null ? row.name() : row.description(),
                    row.category(),
                    row.stepKey(),
                    row.sectionNumber(),
                    row.quantity(),
                    row.unit(),
                    row.unitCost(),
                    row.totalCost(),
                    row.meta()
            ));
        }

        Map<String, Object> bomSpec = new LinkedHashMap<>();
        bomSpec.put("rows", bom.rows());
        bomSpec.put("by_step", bom.byStep());
        bomSpec.put("by_section", bom.bySection());
        bomSpec.put("totals", bom.totals());

        Map<String, Object> pricingSpec = new LinkedHashMap<>();
        pricingSpec.put("lookup", lookup);
        pricingSpec.put("pricing_strategy", pricing);
        pricingSpec.put("schedule", schedule);
        pricingSpec.put("holidays", holidays);
        pricingSpec.put("line_adders", lineAdders == null ? List.of() : lineAdders);
        pricingSpec.put("sections_input", sections);
        pricingSpec.put("quote", quote);
        pricingSpec.put("labour_summary", labour);

        Map<String, Double> totals = new LinkedHashMap<>();
        totals.putAll(quote.totals());
        totals.put("total_cost", quote.totalCost());
        totals.put("target_price", quote.pricing().targetPrice());
        totals.put("rounded_price", quote.pricing().roundedPrice());
        totals.put("actual_profit", quote.pricing().actualProfit());
        totals.put("actual_gm", quote.pricing().actualGm());

        return new CompiledQuotation(configuration, quote, labour, items, bomSpec, pricingSpec, totals);
    }

    public static Map<String, Object> normalizeConfigurationData(Map<String, Object> configData) {
        if (configData == null) {
            return Map.of();
        }

        Map<String, Object> out = new LinkedHashMap<>(configData);
        Object stepLinesRaw = configData.get("stepLines");
        if (!(stepLinesRaw instanceof Map<?, ?> stepLinesAny)) {
            return out;
        }

        @SuppressWarnings("unchecked")
        Map<String, Object> stepLines = (Map<String, Object>) stepLinesAny;

        for (String stepKey : BomEngine.STEP_KEYS) {
            Object linesRaw = stepLines.get(stepKey);
            if (!(linesRaw instanceof List<?> lines) || lines.isEmpty()) {
                continue;
            }

            List<Map<String, Object>> selected = new ArrayList<>();
            for (Object lineRaw : lines) {
                if (!(lineRaw instanceof Map<?, ?> lineAny)) {
                    continue;
                }
                Map<String, Object> line = castMap(lineAny);
                Map<String, Object> mapped = new LinkedHashMap<>();
                mapped.put("component_id", first(line, "componentId", "component_id"));
                mapped.put("part_number", first(line, "partNumber", "part_number"));
                mapped.put("name", line.get("name"));
                mapped.put("unit_cost", first(line, "unitPrice", "unit_cost"));
                Object qty = line.get("quantity");
                mapped.put("quantity", qty instanceof Number ? ((Number) qty).doubleValue() : 1);
                mapped.put("meta", line.containsKey("meta") ? line.get("meta") : Map.of());
                selected.add(mapped);
            }

            Object existing = out.get(stepKey);
            if (existing instanceof Map<?, ?> existingMap) {
                @SuppressWarnings("unchecked")
                Map<String, Object> existingTyped = new LinkedHashMap<>((Map<String, Object>) existingMap);
                boolean alreadyHasData = hasData(existingTyped, "bom_rows")
                        || hasData(existingTyped, "selected_components")
                        || hasData(existingTyped, "items");
                if (!alreadyHasData) {
                    existingTyped.put("selected_components", selected);
                    out.put(stepKey, existingTyped);
                }
            } else {
                out.put(stepKey, Map.of("selected_components", selected));
            }
        }

        return out;
    }

    private static Object first(Map<String, Object> map, String a, String b) {
        if (map.containsKey(a)) {
            return map.get(a);
        }
        return map.get(b);
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> castMap(Map<?, ?> map) {
        return (Map<String, Object>) map;
    }

    private static boolean hasData(Map<String, Object> map, String key) {
        Object raw = map.get(key);
        return raw instanceof List<?> list && !list.isEmpty();
    }
}
